csharp/src/ParallelSelectExt.cs

```diff
@@ -0,0 +1,361 @@
+// This file is part of the ArmoniK project
+//
+// Copyright (C) ANEO, 2022-2023.All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License")
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+using JetBrains.Annotations;
+
+namespace ArmoniK.Utils;
+
+/// <summary>
+///   Options for ParallelSelect and ParallelWait
+/// </summary>
+[PublicAPI]
+public struct ParallelTaskOptions
+{
+  /// <summary>Limit the parallelism for ParallelSelect and ParallelWait</summary>
+  public int ParallelismLimit { get; }
+
+  /// <summary>Cancellation token used for stopping the enumeration</summary>
+  public CancellationToken CancellationToken { get; }
+
+  /// <summary>
+  ///   Options for ParallelSelect and ParallelWait.
+  ///   If parallelismLimit is 0, the number of threads is used as the limit.
+  ///   If parallelismLimit is negative, no limit is enforced.
+  /// </summary>
+  /// <param name="parallelismLimit">Limit the parallelism</param>
+  /// <param name="cancellationToken">Cancellation token used for stopping the enumeration</param>
+  public ParallelTaskOptions(int               parallelismLimit,
+                             CancellationToken cancellationToken = default)
+  {
+    ParallelismLimit = parallelismLimit switch
```

Comment:
As this attribute is computed, we need to test it

Reply:
It is tested indirectly: tests are ran with parallelism limit of -1 and 0, and checked that *effective* parallelism is "infinite" and number of threads.

Reply:
I'm not a fan of the indirect way